include/kernel.h

```diff
@@ -0,0 +1,534 @@
+ *
+ * Threads can receive mails using #kmbox_recv or #kmbox_try_recv.
+ */
+typedef struct kmbox_s
+{
+	/** Read index of the circular buffer */
+	uint8_t r;
+	/** Write index of the circular buffer */
+	uint8_t w;
+	/** Size of the buffer (maximum number of mails) */
+	uint8_t max;
+	/** Number of events this mailbox is attached to */
+	uint8_t evts;
+	/** Linked list of threads waiting on this mailbox */
+	kthread_t *wait_list;
+	/** Circular buffer of mails */
+	void* mails[0];
+} kmbox_t;
+
+/** Maximum number of mailboxes that can be attached to a single event. */
+#define MAX_MAILBOXES_PER_EVENT 8
+
+/**
+ * @brief a kernel event.
+ *
+ * An event, represented by a #kevent_t instance, is a signal that can
+ * be broadcasted to many different threads. A thread can use
+ * #kthread_wait_event to wait for a specific event to fire. The kernel
+ * library includes a set of events bound to MI peripheral interrupts,
+ * so that it's possible for threads to wait for specific interrupts and
+ * act after them.
+ *
+ * It is also possible to attach a mailbox to an event using #mbox_attach_event.
+ * This is useful if a thread needs to wait for multiple events.
+ */
+typedef struct kevent_s
+{
+	/** Mailboxes attached to this event */
+	kmbox_t *mboxes[MAX_MAILBOXES_PER_EVENT];
+} kevent_t;
+
+
+/** @brief Event generated when a SP interrupt is triggered */
+extern kevent_t KEVENT_IRQ_SP;
+/** @brief Event generated when a SI interrupt is triggered */
+extern kevent_t KEVENT_IRQ_SI;
+/** @brief Event generated when a AI interrupt is triggered */
+extern kevent_t KEVENT_IRQ_AI;
+/** @brief Event generated when a VI interrupt is triggered */
+extern kevent_t KEVENT_IRQ_VI;
+/** @brief Event generated when a PI interrupt is triggered */
+extern kevent_t KEVENT_IRQ_PI;
+/** @brief Event generated when a DP interrupt is triggered */
+extern kevent_t KEVENT_IRQ_DP;
+/** @brief Event generated when the RESET button is pushed (pre-NMI exception) */
+extern kevent_t KEVENT_RESET;
+
+
+/** 
+ * @brief Initialize the multi-threading kernel
+ *
+ * The current execution context becomes the main thread of the
+ * program, with priority set to 0 (you can change priority
+ * to the main thread as well using #thread_set_pri).
+ *
+ * The main thread uses the original stack allocated for the
+ * whole process, so it is technically unbounded (or limited by
+ * heap size).
+ *
+ * @return a pointer to the main thread.
+ *
+ */
+kthread_t* kernel_init(void);
+
+/** 
+ * @brief Shutdown the multi-threading kernel
+ *
+ * This function is mostly useful for testing purposes.
+ * Since the kernel does not keep track of all created threads,
+ * this function should be called only when all created threads
+ * are exited or have been killed.
+ */
+void kernel_close(void);
+
+/**
+ * @brief Create a new thread
+ *
+ * Create a new thread, with a specified stack size and priority.
+ * The thread is immediately made ready after creation, so if it has
```

Comment:
Would there be any advantages of having the ability to mark it as not ready yet?

Reply:
The first iteration of the API was actually creating "paused" threads that the user had to unpause. I found it a little overwhelming and without any specific benefit. I think it's fine to create them ready most of the time, since you can always create a thread with a very low priority (so that it's not scheduled even if it's ready) and then later bump it up.